optuna_dashboard/_cached_extra_study_property.py

```diff
@@ -102,10 +105,119 @@ def _update_intermediate_values(self, trial: FrozenTrial) -> None:
             self.has_intermediate_values = True
 
     def _update_search_space(self, trial: FrozenTrial) -> None:
-        current = set([(n, d) for n, d in trial.distributions.items()])
-        self._union_search_space = self._union_search_space.union(current)
+        def _is_same_float_distribution(d1: BaseDistribution, d2: BaseDistribution) -> bool:
+            return (
+                isinstance(d1, FloatDistribution)
+                and isinstance(d2, FloatDistribution)
+                and d1.step == d2.step
+                and d1.log == d2.log
+            )
+
+        def _is_same_int_distribution(d1: BaseDistribution, d2: BaseDistribution) -> bool:
+            return (
+                isinstance(d1, IntDistribution)
+                and isinstance(d2, IntDistribution)
+                and d1.step == d2.step
+                and d1.log == d2.log
+            )
+
+        def _is_same_categorical_distribution(d1: BaseDistribution, d2: BaseDistribution) -> bool:
+            return isinstance(d1, CategoricalDistribution) and isinstance(
+                d2, CategoricalDistribution
+            )
+
+        new_union_search_space: SearchSpaceSetT = set(
+            [(n, d) for n, d in self._union_search_space]
+        )
+        for n1, d1 in trial.distributions.items():
+            for n2, d2 in self._union_search_space:
+                if n1 == n2:
+                    if _is_same_float_distribution(d1, d2):
```

Comment:
I'm just wondering if there is a better approach to mitigate this deep nested code. Do you think we could define like the following function to make the code easier to read?

```python
def get_union_search_space(search_space_1, search_space_1) -> ...:
    ...

def get_intersection_search_space(search_space_1, search_space_1) -> ...:
    ...
```

Reply:
I reduced the nesting by one level, but I couldn't think of a good way to simplify it further...
The functions you mentioned might not be the best since original code requires `break` inside an `if` statement.